core/server/worker/src/main/java/alluxio/worker/block/UfsInputStreamManager.java

```diff
@@ -0,0 +1,300 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.underfs.SeekableUnderFileInputStream;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.IdUtils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalListeners;
+import com.google.common.cache.RemovalNotification;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An under filesystem input stream manager that can cache seekable input streams for future reuse.
+ * The manager caches the input streams, with a time-based eviction policy of expiration configured
+ * in {@link PropertyKey.WORKER_UFS_INSTREAM_CACHE_EXPIRE_MS}.
+ *
+ * An under storage file maybe opened with multiple input streams at the same time. The manager uses
+ * {@link UfsInputStreamIds} to track the in-use input stream and the available ones. The manager
+ * closes the input streams after they are expired and not in-use anymore.
+ */
+@ThreadSafe
+public class UfsInputStreamManager {
+  private static final Logger LOG = LoggerFactory.getLogger(UfsInputStreamManager.class);
+
+  /**
+   * A map from the ufs file name to the metadata of the input streams. Synchronization is needed to
+   * access this map.
+   */
+  private final HashMap<String, UfsInputStreamIds> mFileToInputStreamIds;
+  /**
+   * Cache of the input streams, from the input stream id to the input stream.
+   */
+  @GuardedBy("mFileToInputStreamIds")
+  private final Cache<Long, SeekableUnderFileInputStream> mUnderFileInputStreamCache;
+  /**
+   * Thread pool for asynchronously removing the expired input streams.
+   */
+  private final ExecutorService mRemovalThreadPool;
+
+  /**
+   * Creates a new {@link UfsInputStreamManager}.
+   */
+  public UfsInputStreamManager() {
+    mFileToInputStreamIds = new HashMap<>();
+    mRemovalThreadPool = Executors.newFixedThreadPool(2);
+
+    // A listener to the input stream removal.
+    RemovalListener<Long, SeekableUnderFileInputStream> listener =
```

Comment:
Q: instead of defining a new abstract `SeekableUnderFileInputStream`, can we make it here
`RemovalListener<Long, ? extends InputStream & Seekable>`
Then, you don't need the new class `SeekableUnderFileInputStream` and simply making
`HdfsUnderFileInputStream` extends `FilteredInputStream` and `Seekable` interface.

Reply:
hmm, we still need the resource id and file path for back tracking.

Reply:
`SeekableUnderFileInputStream` seems more like a wrapper class for cache convenience and should be owned by this module as opposed to an `underfs` concept.

Reply:
The challenge is that the under filesystem input stream may use different seekable interface. For example, `RandomAccessFile` defines the `seek` method in the class, while HDFS has `seekable` interface. Ideally, `SeekableUnderFileInputStream ` can be saved if there is a common interface for seeking in the ufs. Since there is no such common abstraction, we need to define on in the `underfs` module.

Reply:
Yeah I agree we should have a wrapper class, ie. `SeekableUnderFileInputStream`, but that should just have the seek api, the additional stuff about ids belongs in another wrapper owned by the ufsinputstreammanager